git/filter_protocol.go

```diff
@@ -0,0 +1,114 @@
+package git
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+)
+
+type protocol struct {
+	r *bufio.Reader
+	w *bufio.Writer
+}
+
+func newProtocolRW(r io.Reader, w io.Writer) *protocol {
+	return &protocol{
+		r: bufio.NewReader(r),
+		w: bufio.NewWriter(w),
+	}
+}
+
+func (p *protocol) readPacket() ([]byte, error) {
+	pktLenHex, err := ioutil.ReadAll(io.LimitReader(p.r, 4))
+	if err != nil || len(pktLenHex) != 4 { // TODO check pktLenHex length
+		return nil, err
+	}
+
+	pktLen, err := strconv.ParseInt(string(pktLenHex), 16, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if pktLen == 0 {
+		return nil, nil
```

Comment:
_Fantastic_ question. This kept me busy most of the morning trying to figure out the best way to handle this sort of thing. The result of my search is summarized in https://github.com/github/git-lfs/pull/1619/commits/9bc1d68b77ca8ad8dc94161c3a74f88e4d462125.

The 🔑  to this being clean is that `0004 <EOF>` is an invalid packet, so the only two cases we have to disambiguate are: 

1. Got a packet with a body (or err)
2. Got a flush (or err)

Here's what I thought:
- Use [`io.ReadFull`](https://github.com/github/git-lfs/blob/9bc1d68b77ca8ad8dc94161c3a74f88e4d462125/git/filter_protocol.go#L42) to wait until we have a _full_ header present on the wire before continuing. This means that if we call `readPacket()` and Git hasn't sent us any data over the pipe yet, the call will block (simplifying the call-sites and removing the need to spin waiting for a packet) instead of return early. This is, IMHO, great behavior, since we don't need to add a nth distinction to the return values of this method where we read a partial packet or didn't have any packet at all.
- Return early if we got less bytes than we wanted, meaning the pipe closed with invalid/partial data left on the wire. This happens when `io.ReadFull` encounters an error, halting the read, we can inspect the data that it slurped up, and exit early.
- Utilize the fact that "empty" packets (those beginning with `0004` and no body) are invalid to disambiguate between invalid packets `(nil, err)`, and flush packets `(nil, nil)`.

To summarize what's in the documentation, this method now has four return scenarios:

1. `([]byte, nil)` - an entire packet was read successfully
2. `(nil, io.EOF)` - no more data, the pipe was closed
3.  `(nil, io.ErrShortBuffer)` - partial data was read, but the pipe was closed before the packet's entire data was pushed onto the pipe
4. `(nil, nil)` - empty body (a flush packet) and no error.

(and the last case, which is nil , and some random error if the header couldn't be parsed, or etc.)